de.metas.contracts/src/main/java/de/metas/contracts/impl/FlatrateBL.java

```diff
@@ -1033,6 +1039,12 @@ public void extendContract(final @NonNull ContractExtendingRequest context)
 				nextTransition = nextConditions.getC_Flatrate_Transition();
 				Check.assumeNotNull(nextTransition, "C_Flatrate_Transition shall not be null!");
 
+				// infinete loop detection
+				if (X_C_Flatrate_Transition.EXTENSIONTYPE_ExtendAll.equals(nextTransition.getExtensionType()) && !conditionsBreadCrumb.add(nextConditions.getC_Flatrate_Conditions_ID()))
+				{
+					throw new AdempiereException("Infinite loop detected!");
```

Comment:
add this in AD_Message.
More, when collecting the seen flatrate conditions, also collect their names.
In case a loop is detected we also want to know why were those conditions that caused the loop.

Reply:
Yes; I am in progress of adding it.
Yep; Good point. I'll add the name also! Thx!